docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,189 @@
+is inconvenient, specially if there are many hosts or if rebooting a host
+requires manual intervention. This is a proposal to implement a mechanism to
+upgrade the agent without having to reboot the host.
+
+## Motivation
+
+Reduce the need to reboot hosts when there are changes in the service that
+require new versions of the agent.
+
+### Goals
+
+A host that has been booted and registered with matching versions of the agent
+and the service will not need to be rebooted in order to install the cluster
+with a newer version of the service.
+
+Note that in this context _version_ is really a container image name, like
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`.
+
+### Non-Goals
+
+Hosts will still need to be rebooted with a new version of the ISO if there
+are changes in the part of the agent that performs the initial registration.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then upgrade the service to a version that
+requires a newer version of the agent, and finally install the cluster without
+having to reboot the hosts.
+
+#### Story 2
+
+As a cluster Creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then downgrade the service to a version
+that requires an older version of the agent, and finally install the cluster
+without having to reboot the hosts.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
+of the `AGENT_DOCKER_IMAGE` environment variable with the value of the
+`discovery_agent_version` stored in the `hosts` table.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, as
+that renders the comparison useless. In the _SaaS_ stage environment that
+variable is already using a tag, so that shouldn't be a problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
```

Comment:
The `upgrade-agent` step is also (similar to register) something we must not break in order to allow agent upgrade.

Reply:
Done.